daemon/logging.c

```diff
@@ -0,0 +1,134 @@
+ * HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * The views and conclusions contained in the software and documentation
+ * are those of the authors and should not be interpreted as representing
+ * any official policies, either expressed or implied.
+ */
+
+#include "logging.h"
+#include "util.h"
+
+#include <stdarg.h>
+#include <errno.h>
+
+/* ========================================================================= *
+ * Prototypes
+ * ========================================================================= */
+
+/* ------------------------------------------------------------------------- *
+ * LOG
+ * ------------------------------------------------------------------------- */
+
+static int         log_normalize_level(int lev);
+static const char *log_tag            (int lev);
+int                log_get_level      (void);
+void               log_set_level      (int lev);
+bool               log_p              (int lev);
+void               log_emit_real      (const char *file, int line, const char *func, int lev, const char *fmt, ...);
+
+/* ========================================================================= *
+ * LOG
+ * ========================================================================= */
+
+static int log_level = LOG_WARNING;
+
+static int
+log_normalize_level(int lev)
+{
+    return (lev < LOG_CRIT) ? LOG_CRIT : (lev > LOG_TRACE) ? LOG_TRACE : lev;
+}
+
+static const char *
+log_tag(int lev)
+{
+    static const char * const tag[] = {
+        [LOG_EMERG]   = "X: ",
+        [LOG_ALERT]   = "A: ",
+        [LOG_CRIT]    = "C: ",
+        [LOG_ERR]     = "E: ",
+        [LOG_WARNING] = "W: ",
+        [LOG_NOTICE]  = "N: ",
+        [LOG_INFO]    = "I: ",
+        [LOG_DEBUG]   = "D: ",
+        [LOG_TRACE]   = "T: ",
+    };
+    return tag[lev];
+}
+
+int
+log_get_level(void)
+{
+    return log_level;
+}
+
+void
+log_set_level(int lev)
+{
+    log_level = log_normalize_level(lev);
+}
+
+bool
+log_p(int lev)
+{
+    return log_normalize_level(lev) <= log_level;
+}
+
+void
+log_emit_real(const char *file, int line, const char *func, int lev, const char *fmt, ...)
+{
+    (void)file;
+    (void)line;
+    (void)func;
+
+    lev = log_normalize_level(lev);
+    if( lev <= log_level ) {
+        int saved = errno;
+
+        char *msg = 0;
+        va_list va;
+        va_start(va, fmt);
+        if( vasprintf(&msg, fmt, va) < 0 )
+            msg = 0;
+        va_end(va);
+
+#if LOGGING_SHOW_FUNCTION
+        fprintf(stderr, "%s:%d: %s(): %s%s\n",
```

Comment:
It is missing use syslog() instead of stderr step. [Personally I see little value in using sd_journal_xxx()  just because such thing exists]

Reply:
Made it so that one can choose stderr or syslog, and --systemd implies syslog -> when run as systemd service uses syslog, when executed from shell logs to stderr.